langchain4j/src/main/java/dev/langchain4j/data/document/splitter/DocumentByMarkdownHeaderSplitter.java

```diff
@@ -0,0 +1,159 @@
+package dev.langchain4j.data.document.splitter;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import dev.langchain4j.data.document.DocumentSplitter;
+import dev.langchain4j.model.Tokenizer;
+
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+
+
+public class DocumentByMarkdownHeaderSplitter extends HierarchicalDocumentSplitter {
+
+    private final String joinDelimiter;
+    private final Pattern pattern;
+
+    /**
+     * Constructor for DocumentByMarkdownHeaderSplitter.
+     *
+     * @param headerLevels          The levels of headers to split by.
+     * @param joinDelimiter         The delimiter to use when joining segments.
+     * @param maxSegmentSizeInChars The maximum size of a segment in characters.
+     * @param maxOverlapSizeInChars The maximum size of overlap between segments in characters.
+     */
+    public DocumentByMarkdownHeaderSplitter(
+            Set<MarkdownHeaderLevel> headerLevels,
+            String joinDelimiter,
+            int maxSegmentSizeInChars,
+            int maxOverlapSizeInChars) {
+        super(maxSegmentSizeInChars, maxOverlapSizeInChars, null, null);
+        this.joinDelimiter = ensureNotNull(joinDelimiter, "joinDelimiter");
```

Comment:
please check all parameters in all ctors for consistency. For example, `joinDelimiter` is mandatory in this ctor, but not in others.

Reply:
Do we want to add also nullability (jakarta.annotation.Nullable/Nonnull)annotations to all parameters and responses? NPE-safe languages like Kotlin and IDE will have good use of it.

Reply:
Feel free to open an issue for this, but let's address it a bit later